# CUNY DATA612 Summer 2020 - Project 5

# John K. Hancock

### Project Scope
The goal of this project is give you practice beginning to work with a distributed recommender system. It is sufficient for this assignment to build out your application on a single node.
Adapt one of your recommendation systems to work with Apache Spark and compare the performance with your previous iteration. Consider the efficiency of the system and the added complexity of using Spark. You may complete the assignment using PySpark (Python), SparkR (R) , sparklyr (R), or Scala. Please include in your conclusion: For your given recommender system’s data, algorithm(s), and (envisioned) implementation, at what point would you see moving to a distributed platform such as Spark becoming necessary?

### Project Plan

For this project, I switched to Python, specifically Apache Spark using the PySpark libraries in order to learn something new. This project was a real stretch for me. I chose to build the movie recommender on the distributed system using Databricks. Additionally, I created a model using Alternating Least Squares ("ALS"). The project begins with a discussion of ALS. Next, I prepare and explore the data. I then split the data into test and train, discuss the ALS parameters and hyper-parameters, fit the model, make predicitions, and finally evaluate the model.

##### Citation: This project is based on the DataCamp class: "Building Recommendation Engines with PySpark" by Jamen Long
###### https://learn.datacamp.com/courses/recommendation-engines-in-pyspark

### Alternating Least Squares (ALS)

Commonoly used in Recommender systems, Collaborative based filtering models are used to find users who share similar interests in order to recommend products to users. User-User Collaborative filtering finds users that are similar to the target user. Item-Item Collaborative filtering finds and recommends items that are similar to items with the target user. Normally, we track user similarities by user ratings matrix where users are the rows and the items are the columns. These user ratings are recorded in a matrix, and most often the matrix will be sparse as most users wont rate most products.

Enter Alternating Least Squares ("ALS"). ALS uses a factorization method called non-negative factorization. ALS factors a matrix into 2 matrixes starting with random numbers. It then re-populated the two matrixes unti it approximates the original matrix. It holds the original matrix and the first factored matrix constant while it makes adjustments to the second factored matrix.  In other factorizations, the matrixes can return negative values after factorization. ALS requires that factorizations only return positive values. The dimensions of the factor matrixes that don't match the original matrix are called the Rank or Latent features. We can choose the rank or the number of latent features. The Root Mean Squared Error determines how close the factored matrixes are to the original matrix. Through iterations, ALS adjusts both the first and second matrix until the RMSE is minimized. The end result is a dense matrix which can be used to fill in the missing values of the sparse matrix.

### Load the Libraries

In [6]:
#Import the pyspark sql libraries.  These are used to execute sql queries.
from pyspark import SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import SQLContext
from pyspark.sql.functions import col, avg, min

#Import the pyspark machine learning libraries
from pyspark.ml.recommendation import ALS
from pyspark.ml.evaluation import RegressionEvaluator

from pyspark.ml.tuning import ParamGridBuilder, CrossValidator
from pyspark.mllib.recommendation import MatrixFactorizationModel, Rating
from pyspark.mllib.classification import LogisticRegressionWithLBFGS
from pyspark.mllib.evaluation import BinaryClassificationMetrics


sqlContext = SQLContext(sc)

import numpy as np
import pandas as pd

### Data Preparation and Exploration
In this section, I loaded the GroupLens MovieLens ratings dataset. Using the textFile method on the Spark context object, "sc", the data is read into Spark's primary data structure, Resilient Distributed Dataset (RDD) =  Resilient, fault tolerant; Distributed, data resides on multiple nodes; Dataset, the records of the dataset.

In [8]:
data = sc.textFile("/FileStore/tables/ratings.csv")
data.take(5)

Out[2]: ['userId,movieId,rating,timestamp',
 '1,1,4.0,964982703',
 '1,3,4.0,964981247',
 '1,6,4.0,964982224',
 '1,47,5.0,964983815']

I then converted the RDD into a spark data frame using the Spark SQL method read.csv method.

In [10]:
ratings = spark.read.csv(data, header=True, inferSchema=True)
ratings.take(5)

Out[3]: [Row(userId=1, movieId=1, rating=4.0, timestamp=964982703),
 Row(userId=1, movieId=3, rating=4.0, timestamp=964981247),
 Row(userId=1, movieId=6, rating=4.0, timestamp=964982224),
 Row(userId=1, movieId=47, rating=5.0, timestamp=964983815),
 Row(userId=1, movieId=50, rating=5.0, timestamp=964982931)]

The columns of the dataset are 'userIsd', 'movieId', 'rating', and 'timestamp'.  Over the next two blocks, we see the first 20 columns of the dataset along with the schema.

In [12]:
print(ratings.columns)

['userId', 'movieId', 'rating', 'timestamp']

In [13]:
print(ratings.show())

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
+------+-------+------+---------+
only showing top 20 rows

None

In [14]:
ratings

Out[6]: DataFrame[userId: int, movieId: int, rating: double, timestamp: int]

The following block computes the sparsity of the ratings matrix which shows 98.3% empty

In [16]:
# Count the total number of ratings in the dataset
numerator = ratings.select("rating").count()

# Count the number of distinct userIds and distinct movieIds
num_users = ratings.select("userId").distinct().count()
num_movies = ratings.select("movieId").distinct().count()
denominator = num_users * num_movies
sparsity = (1.0 - (numerator *1.0)/denominator)*100
print("The ratings dataframe is ", "%.2f" % sparsity + "% empty.")

The ratings dataframe is 98.30% empty.

In [17]:
ratings.select("rating").count()

Out[8]: 100836

Below we see the first 20 users in the dataset

In [19]:
ratings.filter(col("userId") < 100).show()

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
+------+-------+------+---------+
only showing top 20 rows

##### Users with the most ratings

In [21]:
print("Users with the most ratings: ")
display(ratings.groupBy("userId").count().sort(col("count").desc()))

userId count 414 2698 599 2478 474 2108 448 1864 274 1346 610 1302 68 1260 380 1218 606 1115 288 1055 249 1046 387 1027 182 977 307 975 603 943 298 939 177 904 318 879 232 862 480 836 608 831 600 763 590 728 483 728 105 722 19 703 305 677 489 648 111 646 438 635 217 613 140 608 477 600 555 578 91 575 28 570 219 528 534 520 89 518 64 517 226 507 561 505 18 502 525 500 57 476 381 474 368 469 509 467 469 465 560 458 462 455 292 446 597 443 21 443 42 440 160 437 294 437 580 436 596 411 202 403 275 403 517 400 45 399 156 398 514 397 391 386 567 385 357 383 103 377 339 371 62 366 199 363 125 360 51 359 132 347 66 345 313 340 200 334 221 331 6 314 453 311 50 310 425 306 428 300 573 299 352 294 84 293 122 292 382 291 356 289 239 279 135 279 365 277 484 275 104 273 63 271 325 270 169 269 332 267 290 267 495 265 432 260 187 258 328 255 222 250 212 248 330 247 20 242 282 237 372 236 434 233 1 232 594 232 562 231 198 230 82 227 354 226 186 226 312 223 605 221 41 217 304 216 4 216 119 215 563 213 73 210 263 210 586 208 220 207 246 204 452 202 234 202 385 201 522 200 280 196 139 194 570 191 166 190 317 189 552 188 195 187 607 187 520 186 367 185 479 181 266 180 153 179 74 177 279 176 256 174 167 173 331 168 95 168 141 168 527 167 80 167 410 167 117 165 587 165 216 164 346 164 265 163 577 161 564 157 33 156 475 155 572 155 419 155 334 154 260 152 7 152 287 152 326 152 247 150 233 150 113 150 100 148 408 141 351 141 137 141 420 141 47 140 129 140 559 140 10 140 230 139 210 138 464 138 424 138 602 135 15 135 27 135 184 134 376 133 377 131 524 131 52 130 268 129 369 129 240 128 314 128 482 128 144 128 411 127 109 127 254 127 409 126 551 123 393 123 437 122 492 121 23 121 405 120 22 119 76 119 181 118 83 118 308 115 465 115 43 114 301 114 466 113 542 113 490 113 58 112 115 112 306 112 571 112 136 111 488 111 24 110 201 110 362 109 510 108 322 107 59 107 436 106 286 106 17 105 40 103 593 103 309 103 32 102 412 102 601 101 604 100 39 100 373 100 323 98 215 98 16 98 361 97 93 97 159 97 503 95 168 94 227 94 415 94 592 94 370 93 244 93 422 93 418 93 98 92 353 90 554 89 211 89 284 88 426 88 541 87 504 87 116 87 500 86 34 86 191 85 213 84 446 84 427 84 553 83 204 83 584 83 171 82 460 82 29 81 390 81 470 80 38 78 96 78 447 78 337 77 178 77 445 77 241 76 108 76 543 76 223 75 225 75 523 75 566 74 359 74 528 73 579 73 297 72 401 72 143 71 86 70 384 70 75 69 476 69 131 69 344 69 179 69 386 68 417 67 174 67 190 66 342 66 235 65 112 65 165 65 229 65 11 64 546 64 491 64 79 64 310 64 152 63 70 62 345 62 101 61 585 61 395 61 78 61 493 61 36 60 404 60 458 59 151 59 343 59 121 58 430 58 526 58 429 58 149 58 267 57 455 57 183 57 262 57 264 56 558 56 588 56 94 56 582 56 102 56 486 56 321 56 487 56 88 56 413 56 123 56 583 56 336 56 273 55 348 55 224 54 591 54 90 54 261 53 303 53 316 53 99 53 341 52 511 52 450 51 497 51 416 51 110 51 248 51 532 50 457 50 170 50 512 50 124 50 31 50 237 50 402 48 188 48 14 48 148 48 327 48 378 48 44 48 8 47 253 47 537 47 185 47 56 46 69 46 155 46 398 46 9 46 203 45 347 45 72 45 501 45 454 45 441 45 238 44 255 44 5 44 400 43 271 43 456 43 506 42 540 42 444 42 435 42 46 42 358 41 295 41 276 41 371 41 536 40 581 40 350 40 521 40 589 40 270 40 338 39 161 39 61 39 3 39 142 38 126 38 449 38 162 38 539 38 533 38 252 38 473 37 443 37 349 37 538 37 609 37 421 37 379 36 243 36 97 36 399 36 176 36 164 36 285 36 575 36 67 36 209 35 283 35 319 35 133 35 193 35 502 35 242 35 498 35 197 35 71 35 134 35 383 34 389 34 403 34 65 34 154 34 85 34 107 34 451 34 30 34 54 33 463 33 374 33 468 33 375 33 48 33 128 33 106 33 315 33 440 33 12 32 556 32 302 32 513 32 300 32 146 32 196 32 291 31 481 31 505 31 13 31 114 31 272 31 366 31 363 30 535 30 236 30 388 29 77 29 269 29 565 29 496 29 259 29 2 29 472 29 396 28 550 28 335 28 471 28 130 28 277 28 340 28 311 28 205 27 250 27 296 27 530 27 461 27 499 27 578 27 557 27 150 26 158 26 516 26 218 26 208 26 25 26 355 26 515 26 81 26 172 26 548 26 459 26 519 26 529 25 360 25 333 25 206 25 392 25 55 25 173 25 258 25 228 2

In [22]:
print("Users with the most ratings.")
ratings.groupBy("userId").count().sort(col("count").desc()).show()


Users with the most ratings.
+------+-----+
userId|count|
+------+-----+
 414| 2698|
 599| 2478|
 474| 2108|
 448| 1864|
 274| 1346|
 610| 1302|
 68| 1260|
 380| 1218|
 606| 1115|
 288| 1055|
 249| 1046|
 387| 1027|
 182| 977|
 307| 975|
 603| 943|
 298| 939|
 177| 904|
 318| 879|
 232| 862|
 480| 836|
+------+-----+
only showing top 20 rows

In [23]:
print("Users with the fewest ratings: ")
ratings.groupBy("userId").count().sort(col("count").asc()).show()

Users with the fewest ratings: 
+------+-----+
userId|count|
+------+-----+
 576| 20|
 569| 20|
 320| 20|
 595| 20|
 53| 20|
 431| 20|
 257| 20|
 442| 20|
 189| 20|
 194| 20|
 278| 20|
 207| 20|
 406| 20|
 147| 20|
 507| 21|
 87| 21|
 439| 21|
 549| 21|
 598| 21|
 324| 21|
+------+-----+
only showing top 20 rows

##### Users with the Fewest Ratings

In [25]:
print("Users with the fewest ratings: ")
display(ratings.groupBy("userId").count().sort(col("count").asc()))

userId count 406 20 431 20 194 20 53 20 569 20 320 20 576 20 257 20 595 20 189 20 207 20 147 20 442 20 278 20 549 21 49 21 87 21 439 21 507 21 293 21 157 21 364 21 245 21 598 21 26 21 37 21 547 21 324 21 281 21 467 22 544 22 120 22 214 22 433 22 60 22 478 22 127 22 531 22 118 22 192 22 138 22 494 22 407 22 545 23 163 23 394 23 397 23 299 23 145 23 574 23 423 23 35 23 329 23 251 23 568 23 485 23 518 24 289 24 92 24 508 24 175 24 231 24 180 24 228 25 55 25 392 25 258 25 360 25 529 25 206 25 333 25 173 25 208 26 516 26 548 26 459 26 81 26 25 26 519 26 515 26 355 26 150 26 218 26 158 26 172 26 578 27 250 27 530 27 296 27 461 27 499 27 557 27 205 27 340 28 130 28 277 28 396 28 550 28 471 28 335 28 311 28 565 29 388 29 77 29 472 29 496 29 269 29 2 29 259 29 236 30 363 30 535 30 272 31 481 31 366 31 291 31 13 31 505 31 114 31 300 32 556 32 12 32 513 32 146 32 302 32 196 32 468 33 128 33 48 33 54 33 463 33 106 33 375 33 440 33 374 33 315 33 65 34 403 34 389 34 154 34 383 34 451 34 107 34 85 34 30 34 283 35 502 35 242 35 209 35 133 35 193 35 319 35 71 35 197 35 498 35 134 35 164 36 176 36 97 36 285 36 243 36 575 36 379 36 67 36 399 36 609 37 443 37 421 37 349 37 538 37 473 37 449 38 162 38 533 38 539 38 252 38 126 38 142 38 338 39 161 39 61 39 3 39 521 40 589 40 536 40 350 40 270 40 581 40 295 41 371 41 358 41 276 41 540 42 444 42 506 42 435 42 46 42 271 43 400 43 456 43 238 44 255 44 5 44 454 45 501 45 347 45 203 45 72 45 441 45 155 46 69 46 56 46 9 46 398 46 8 47 537 47 185 47 253 47 44 48 402 48 148 48 378 48 327 48 188 48 14 48 31 50 512 50 532 50 237 50 457 50 170 50 124 50 450 51 416 51 110 51 248 51 497 51 511 52 341 52 261 53 303 53 99 53 316 53 591 54 224 54 90 54 348 55 273 55 94 56 264 56 487 56 88 56 321 56 588 56 582 56 102 56 558 56 336 56 413 56 486 56 583 56 123 56 267 57 183 57 262 57 455 57 430 58 149 58 429 58 121 58 526 58 151 59 343 59 458 59 36 60 404 60 395 61 585 61 493 61 101 61 78 61 345 62 70 62 152 63 491 64 79 64 11 64 546 64 310 64 165 65 229 65 235 65 112 65 190 66 342 66 417 67 174 67 386 68 344 69 131 69 476 69 75 69 179 69 384 70 86 70 143 71 297 72 401 72 528 73 579 73 566 74 359 74 223 75 225 75 523 75 543 76 241 76 108 76 337 77 445 77 178 77 96 78 38 78 447 78 470 80 390 81 29 81 171 82 460 82 553 83 204 83 584 83 446 84 213 84 427 84 191 85 500 86 34 86 116 87 541 87 504 87 426 88 284 88 554 89 211 89 353 90 98 92 244 93 418 93 422 93 370 93 168 94 227 94 415 94 592 94 503 95 361 97 93 97 159 97 16 98 215 98 323 98 373 100 604 100 39 100 601 101 32 102 412 102 309 103 593 103 40 103 17 105 436 106 286 106 322 107 59 107 510 108 362 109 201 110 24 110 136 111 488 111 571 112 306 112 58 112 115 112 542 113 466 113 490 113 43 114 301 114 465 115 308 115 181 118 83 118 76 119 22 119 405 120 23 121 492 121 437 122 551 123 393 123 409 126 411 127 109 127 254 127 314 128 240 128 144 128 482 128 369 129 268 129 52 130 377 131 524 131 376 133 184 134 27 135 602 135 15 135 424 138 464 138 210 138 230 139 10 140 129 140 559 140 47 140 408 141 137 141 351 141 420 141 100 148 247 150 113 150 233 150 260 152 326 152 7 152 287 152 334 154 475 155 419 155 572 155 33 156 564 157 577 161 265 163 216 164 346 164 587 165 117 165 527 167 410 167 80 167 95 168 331 168 141 168 167 173 256 174 279 176 74 177 153 179 266 180 479 181 367 185 520 186 607 187 195 187 552 188 317 189 166 190 570 191 139 194 280 196 522 200 385 201 452 202 234 202 246 204 220 207 586 208 73 210 263 210 563 213 119 215 4 216 304 216 41 217 605 221 312 223 354 226 186 226 82 227 198 230 562 231 594 232 1 232 434 233 372 236 282 237 20 242 330 247 212 248 222 250 328 255 187 258 432 260 495 265 290 267 332 267 169 269 325 270 63 271 104 273 484 275 365 277 239 279 135 279 356 289 382 291 122 292 84 293 352 294 573 299 428 300 425 306 50 310 453 311 6 314 221 331 200 334 313 340 66 345 132 347 51 359 125 360 199 363 62 366 339 371 103 377 357 383 567 385 391 386 514 397 156 398 45 399 517 400 202 403 275 403 596 411 580 436 294 437 160 437 42 440 

In [26]:
print("Movie with the fewest ratings: ")
ratings.groupBy("movieId").count().sort(col("count").asc()).show()

Movie with the fewest ratings: 
+-------+-----+
movieId|count|
+-------+-----+
 99130| 1|
 4078| 1|
 6598| 1|
 75341| 1|
 142444| 1|
 8911| 1|
 5155| 1|
 123553| 1|
 26736| 1|
 157775| 1|
 6597| 1|
 111680| 1|
 451| 1|
 8650| 1|
 113829| 1|
 27255| 1|
 137| 1|
 27480| 1|
 34450| 1|
 50610| 1|
+-------+-----+
only showing top 20 rows

##### The average number of ratings per movie is 10.37

In [28]:
# Avg num ratings per movie
print("Avg num ratings per movie: ")
ratings.groupBy("movieId").count().select(avg("count")).show()

Avg num ratings per movie: 
+------------------+
 avg(count)|
+------------------+
10.369806663924312|
+------------------+

##### The fewest rating by a user is 20

In [30]:
print("User with the fewest ratings: ")
ratings.groupBy("userId").count().select(min("count")).show()

User with the fewest ratings: 
+----------+
min(count)|
+----------+
 20|
+----------+

##### The average number of ratings by a user is 165.30

In [32]:
# Avg num ratings per users
ratings.groupBy("userId").count().select(avg("count")).show()

+------------------+
 avg(count)|
+------------------+
165.30491803278687|
+------------------+

### Split Data into train and test

The data is split between train and test using randomSplit.  Additionally, I cached the data as well to ensure faster processing.

In [34]:
# Create test and train set
(train, test) = ratings.randomSplit([0.8, 0.2], seed = 1234)
train.cache()
test.cache()

Out[18]: DataFrame[userId: int, movieId: int, rating: double, timestamp: int]

### Create the ALS model - Parameters and Hyperparameters

userCol, itemCol and ratingCol are the parameters taken from the dataset.

"rank" represents the number of latent features that ALS allows you to set

"maxIter" tells the ALS how many times to adjust the matrixes to reduce the RMSE. Here I set it to 5 to adjust for time

"regParam" is the regularization parameter which is a number added to the error metric to keep it from overfitting

"nonnegative" ensures no negative numbers in the matrix factors

"coldStartStrategy" setting it to drop means preventing users whose data is entirely in the train set from being used to calculate the RMSE

"implicitPrefs" setting it to False means that I want the ALS model to treat the ratings as explicit preferences

In [36]:
als = ALS(userCol="userId", itemCol="movieId", ratingCol="rating", 
          rank =15,
          regParam = .05,
          maxIter = 3,
          nonnegative = True,
          coldStartStrategy="drop",
          implicitPrefs = False)

In [37]:
type(als)

Out[20]: pyspark.ml.recommendation.ALS

Below, I fit the model to the training set

In [39]:
model = als.fit(train)

Below I make predictions on the test set.

In [41]:
# Predict the model  
predictions = model.transform(test)
predictions.take(5)


Out[22]: [Row(userId=436, movieId=471, rating=3.0, timestamp=833530187, prediction=4.352611064910889),
 Row(userId=218, movieId=471, rating=4.0, timestamp=1111624874, prediction=4.03478479385376),
 Row(userId=387, movieId=471, rating=3.0, timestamp=1139047519, prediction=2.826578140258789),
 Row(userId=610, movieId=471, rating=4.0, timestamp=1479544381, prediction=3.3484222888946533),
 Row(userId=217, movieId=471, rating=2.0, timestamp=955943727, prediction=3.289472818374634)]

### Evaluate the Model

Below I used the RegressionEvaluator function to create an evaluator object with the metricName, "rmse"

In [43]:
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")

In [44]:
print(evaluator.getMetricName())
print(evaluator.getLabelCol())
print(evaluator.getPredictionCol())

rmse
rating
prediction

In [45]:
# Calculate and print the RMSE of test_predictions
RMSE = evaluator.evaluate(predictions)
print(RMSE)

0.9116563102818794

### Summary

The RMSE for this model is .911 compared to Assignment 2, the UBCF model resulted in a .99 RMSE.  The ALS model improved on the results.  The overall speed and efficiency of the Apache Spark model is far superior.  Project two's project had 22,112 ratings. This project had 100,836 ratings, yet the entire processing time for this project was lower. In my final project, I will use cross validation to find the best model.